modules/internal/src/main/scala/higherkindness/mu/rpc/internal/server/package.scala

```diff
@@ -56,4 +59,16 @@ package object server {
       }
   }
 
+  private[internal] def extractTracingKernel(headers: Metadata): Kernel = {
+    val asciiHeaders = headers
+      .keys()
+      .asScala
+      .collect {
+        case k if !k.endsWith(BINARY_HEADER_SUFFIX) =>
```

Comment:
dumb question, is it that all internal headers are, if any, binary?

Reply:
gRPC specifies that all custom headers must include the `-bin` suffix in their names if the header value is binary, and NOT include the suffix if the value is valid ASCII. So a server knows, just by looking at the header name, how to deserialize the value.

I don't think it's related to whether they are "internal" headers or not.

All the gory details are in the spec: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests

Reply:
Thanks 👍